File: modules/kasane-core/src/document.cpp
// SPDX-License-Identifier: MIT
#include <kasane/document.hpp>
#include <cmath>
#include <unordered_set>

namespace kasane {
bool valid_uuid(const std::string &id) {
    if (id.size() != 36)
        return false;
    bool nonzero = false;
    for (size_t i = 0; i < id.size(); ++i) {
        if (i == 8 || i == 13 || i == 18 || i == 23) {
            if (id[i] != '-')
                return false;
        } else {
            if (!((id[i] >= '0' && id[i] <= '9') || (id[i] >= 'a' && id[i] <= 'f')))
                return false;
            nonzero |= id[i] != '0';
        }
    }
    return nonzero;
}

Status Document::initialize(std::string id, Canvas canvas) {
    if (initialized())
        return Status::error("ALREADY_INITIALIZED", "Create a new Document to open another model.");
    if (!valid_uuid(id))
        return Status::error("INVALID_ID", "Use a nonzero canonical lowercase UUID.");
    if (!std::isfinite(canvas.width) || !std::isfinite(canvas.height) || canvas.width <= 0 ||
        canvas.height <= 0 || !std::isfinite(canvas.origin.x) || !std::isfinite(canvas.origin.y) ||
        !std::isfinite(canvas.pixels_per_unit) || canvas.pixels_per_unit <= 0)
        return Status::error("INVALID_CANVAS", "Canvas dimensions must be finite and positive.");
    id_ = std::move(id);
    canvas_ = canvas;
    return {};
}

bool Document::contains_id(const std::string &id) const {
    return transforms_.contains(id) || parts_.contains(id) || scene_bindings_.contains(id) || id == id_ ||
           assets_.contains(id) || meshes_.contains(id) || deformers_.contains(id) ||
           parameters_.contains(id) || bindings_.contains(id);
}

const ImageAsset *Document::get_asset(const std::string &id) const {
    auto it = assets_.find(id);
    return it == assets_.end() ? nullptr : &it->second;
}

const Mesh *Document::get_mesh(const std::string &id) const {
    auto it = meshes_.find(id);
    return it == meshes_.end() ? nullptr : &it->second;
}

EditResult Document::failed(Status status) const {
    return {std::move(status), {ChangeKind::none, {}, revision_, {}}, {}};
}

void Document::advance_state() {
    current_state_id_ = next_state_id_++;
}

EditResult Document::changed(ChangeKind kind, std::vector<std::string> ids,
                             std::vector<std::string> objects) {
    if (objects.empty())
        objects = ids;
    advance_state();
    return {{}, {kind, std::move(ids), ++revision_, std::move(objects)}, {}};
}

EditResult Document::add_asset(ImageAsset asset) {
    if (mutation_blocked())
        return failed(Status::error("TRANSACTION_ACTIVE", "Commit or cancel the active transaction first."));
    if (!initialized())
        return failed(Status::error("NOT_INITIALIZED", "Initialize Document first."));
    if (!valid_uuid(asset.id))
        return failed(Status::error("INVALID_ID", "Asset ID must be a canonical UUID."));
    if (contains_id(asset.id))
        return failed(Status::error("DUPLICATE_ID", "Object ID already exists."));
    if (!asset.width || !asset.height || asset.source.empty())
        return failed(Status::error("INVALID_ASSET", "Asset requires source and positive pixel dimensions."));
    const auto key = asset.id;
    assets_.emplace(key, std::move(asset));
    asset_order_.push_back(key);
    return changed(ChangeKind::metadata, {}, {key});
}

EditResult Document::create_mesh(Mesh mesh) {
    if (mutation_blocked())
        return failed(Status::error("TRANSACTION_ACTIVE", "Commit or cancel the active transaction first."));
    if (!initialized())
        return failed(Status::error("NOT_INITIALIZED", "Initialize Document first."));
    if (!valid_uuid(mesh.id))
        return failed(Status::error("INVALID_ID", "Mesh ID must be a canonical UUID."));
    if (contains_id(mesh.id))
        return failed(Status::error("DUPLICATE_ID", "Object ID already exists."));
    if (!get_asset(mesh.texture_asset_id))
        return failed(Status::error("MISSING_ASSET", "Texture asset does not exist."));
    if (auto s = validate_mesh_properties(mesh); !s.ok())
        return failed(s);
    if (mesh.runtime_id.empty())
        mesh.runtime_id = mesh.id;
    for (const auto &[id, other] : meshes_)
        if (other.runtime_id == mesh.runtime_id)
            return failed(Status::error("DUPLICATE_RUNTIME_ID", mesh.id + ".runtime_id duplicates " + id));
    if (mesh.vertex_ids.size() != mesh.base_positions.size())
        return failed(Status::error("INVALID_LENGTH", "Vertex IDs must match positions."));
    std::unordered_map<VertexId, uint32_t> slots;
    for (size_t i = 0; i < mesh.vertex_ids.size(); ++i) {
        if (!slots.emplace(mesh.vertex_ids[i], static_cast<uint32_t>(i)).second)
            return failed(Status::error("DUPLICATE_VERTEX", "Vertex IDs must be unique within a mesh."));
    }
    std::vector<uint32_t> indices;
    for (const auto &triangle : mesh.triangles) {
        for (auto vertex : triangle) {
            auto it = slots.find(vertex);
            if (it == slots.end())
                return failed(Status::error("MISSING_VERTEX", "Triangle references an unknown vertex ID."));
            indices.push_back(it->second);
        }
    }
    if (auto status = validate_render_mesh(mesh.base_positions, mesh.uvs, indices); !status.ok())
        return failed(status);
    const auto key = mesh.id;
    meshes_.emplace(key, std::move(mesh));
    vertex_slots_.emplace(key, std::move(slots));
    mesh_order_.push_back(key);
    return changed(ChangeKind::structure, {key});
}

EditResult Document::rename_mesh(const std::string &id, std::string name) {
    if (mutation_blocked())
        return failed(Status::error("TRANSACTION_ACTIVE", "Commit or cancel the active transaction first."));
    auto it = meshes_.find(id);
    if (it == meshes_.end())
        return failed(Status::error("MISSING_MESH", "Mesh does not exist."));
    if (it->second.name == name)
        return {{}, {ChangeKind::none, {}, revision_, {}}, {}};
    it->second.name = std::move(name);
    return changed(ChangeKind::metadata, {id});
}

EditResult Document::set_vertex_positions(const std::string &id, std::span<const VertexId> vertices,
                                          std::span<const Vec2> positions) {
    VertexPositionUpdate update{id, {vertices.begin(), vertices.end()}, {positions.begin(), positions.end()}};
    return apply_vertex_position_updates(std::span<const VertexPositionUpdate>(&update, 1));
}

EditResult Document::apply_vertex_position_updates(std::span<const VertexPositionUpdate> updates) {
    if (mutation_blocked())
        return failed(Status::error("TRANSACTION_ACTIVE", "Use commit_transaction for staged edits."));

    struct PositionDelta {
        std::string mesh_id;
        std::vector<uint32_t> slots;
        std::vector<Vec2> after;
    };

    std::vector<PositionDelta> deltas;
    std::unordered_map<std::string, std::unordered_set<VertexId>> seen;
    std::vector<std::string> changed_meshes;
    std::unordered_set<std::string> changed_mesh_set;
    for (const auto &update : updates) {
        auto mesh_it = meshes_.find(update.mesh_id);
        if (mesh_it == meshes_.end())
            return failed(Status::error("MISSING_MESH", "Mesh does not exist."));
        if (update.vertex_ids.size() != update.positions.size())
            return failed(Status::error("INVALID_LENGTH", "IDs and positions must match."));
        if (auto status = validate_positions(update.positions); !status.ok())
            return failed(status);
        PositionDelta delta;
        delta.mesh_id = update.mesh_id;
        const auto &lookup = vertex_slots_.at(update.mesh_id);
        for (size_t i = 0; i < update.vertex_ids.size(); ++i) {
            const auto vertex = update.vertex_ids[i];
            if (!seen[update.mesh_id].insert(vertex).second)
                return failed(
                    Status::error("DUPLICATE_VERTEX", "A transaction cannot write a vertex twice."));
            auto slot = lookup.find(vertex);
            if (slot == lookup.end())
                return failed(Status::error("MISSING_VERTEX", "Vertex ID does not exist."));
            const auto old = mesh_it->second.base_positions[slot->second];
            if (old == update.positions[i])
                continue;
            delta.slots.push_back(slot->second);
            delta.after.push_back(update.positions[i]);
        }
        if (!delta.slots.empty()) {
            if (changed_mesh_set.insert(update.mesh_id).second)
                changed_meshes.push_back(update.mesh_id);
            deltas.push_back(std::move(delta));
        }
    }
    if (deltas.empty())
        return {{}, {ChangeKind::none, {}, revision_, {}}, {}};
    for (const auto &delta : deltas) {
        auto &positions = meshes_.at(delta.mesh_id).base_positions;
        for (size_t i = 0; i < delta.slots.size(); ++i)
            positions[delta.slots[i]] = delta.after[i];
    }
    return changed(ChangeKind::positions, std::move(changed_meshes));
}

EditResult Document::apply_vertex_position_updates_at_revision(std::span<const VertexPositionUpdate> updates,
                                                               uint64_t expected_revision) {
    if (revision_ != expected_revision)
        return failed(Status::error("STALE_REVISION", "Document changed since this transaction began."));
    return apply_vertex_position_updates(updates);
}

Status Document::begin_transaction() {
    if (!initialized())
        return Status::error("NOT_INITIALIZED", "Initialize Document first.");
    if (transaction_active_)
        return Status::error("TRANSACTION_ACTIVE", "A transaction is already active.");
    transaction_active_ = true;
    staged_updates_.clear();
    return {};
}

Status Document::stage_vertex_positions(VertexPositionUpdate update) {
    if (!transaction_active_)
        return Status::error("NO_TRANSACTION", "Begin a transaction first.");
    if (update.vertex_ids.size() != update.positions.size())
        return Status::error("INVALID_LENGTH", "IDs and positions must match.");
    if (auto status = validate_positions(update.positions); !status.ok())
        return status;
    if (!get_mesh(update.mesh_id))
        return Status::error("MISSING_MESH", "Mesh does not exist.");
    staged_updates_.push_back(std::move(update));
    return {};
}

EditResult Document::commit_transaction() {
    if (!transaction_active_)
        return failed(Status::error("NO_TRANSACTION", "Begin a transaction first."));
    transaction_active_ = false;
    auto updates = std::move(staged_updates_);
    staged_updates_.clear();
    return apply_vertex_position_updates(updates);
}

Status Document::cancel_transaction() {
    if (!transaction_active_)
        return Status::error("NO_TRANSACTION", "No transaction is active.");
    staged_updates_.clear();
    transaction_active_ = false;
    return {};
}

void Document::restore_from(const Document &source) {
    const auto next_revision = revision_ + 1;
    const auto next_state = next_state_id_;
    *this = source;
    revision_ = next_revision;
    next_state_id_ = next_state;
    advance_state();
    transaction_active_ = false;
    staged_updates_.clear();
}

EditResult Document::replace_mesh(Mesh mesh) {
    if (mutation_blocked())
        return failed(Status::error("TRANSACTION_ACTIVE", "Commit or cancel the active transaction first."));
    if (!get_mesh(mesh.id))
        return failed(Status::error("MISSING_MESH", "Mesh does not exist."));
    const auto *previous = get_mesh(mesh.id);
    if (binding_for_mesh(mesh.id) &&
        (previous->vertex_ids != mesh.vertex_ids || previous->triangles != mesh.triangles))
        return failed(
            Status::error("KEYFORMS_REQUIRED",
                          mesh.id + ": topology replacement requires every keyform and vertex mapping"));
    // Validate through the same creation path before replacing any live arrays.
    Document candidate = *this;
    candidate.meshes_.erase(mesh.id);
    candidate.vertex_slots_.erase(mesh.id);
    const auto key = mesh.id;
    auto edit = candidate.create_mesh(std::move(mesh));
    if (!edit.status.ok())
        return failed(edit.status);
    meshes_.at(key) = std::move(candidate.meshes_.at(key));
    vertex_slots_.at(key) = std::move(candidate.vertex_slots_.at(key));
    return changed(ChangeKind::structure, {key});
}

Status Document::render_indices(const std::string &id, std::vector<uint32_t> &out) const {
    const auto *mesh = get_mesh(id);
    if (!mesh)
        return Status::error("MISSING_MESH", "Mesh does not exist.");
    const auto &slots = vertex_slots_.at(id);
    std::vector<uint32_t> next;
    next.reserve(mesh->triangles.size() * 3);
    for (const auto &triangle : mesh->triangles)
        for (auto vertex : triangle)
            next.push_back(slots.at(vertex));
    out = std::move(next);
    return {};
}
} // namespace kasane

File: modules/kasane-core/src/document.rs
mod assets;
mod bindings;
mod blendshapes;
mod canvas;
mod checkpoint;
mod glue;
mod meshes;
mod offscreen;
mod parameters;
mod parts;
mod references;
mod structure;
mod transactions;
mod transforms;
mod validation;
mod vertices;

pub use checkpoint::DocumentCheckpoint;
pub use structure::StructureIssue;
pub use validation::*;

use crate::draw_order::DrawOrderGroup;
use crate::types::{
    BlendShapeBinding, BlendShapeConstraint, BlendShapeKeyTable, Canvas, ChangeKind, ChangeSet,
    DocumentEdit, EditResult, Glue, ImageAsset, Mesh, MeshBinding, Offscreen, Parameter, Part,
    SceneBinding, Status, Transform, VertexId,
};
use std::collections::HashMap;
use std::sync::OnceLock;

pub(super) trait EmptyOr {
    fn empty_or(&self) -> bool;
}

impl<T> EmptyOr for Vec<T> {
    fn empty_or(&self) -> bool {
        self.is_empty()
    }
}

#[derive(Debug, Clone, Default)]
pub struct Document {
    id: String,
    canvas: Canvas,
    revision: u64,
    evaluation_revision: u64,
    batch_build_active: bool,
    transaction_active: bool,
    staged_edits: Vec<DocumentEdit>,
    receipt: crate::history::Receipt,

    assets: HashMap<String, ImageAsset>,
    asset_order: Vec<String>,

    parts: HashMap<String, Part>,
    part_order: Vec<String>,

    transforms: HashMap<String, Transform>,
    transform_order: Vec<String>,

    meshes: HashMap<String, Mesh>,
    vertex_slots: HashMap<String, HashMap<VertexId, usize>>,
    mesh_order: Vec<String>,

    parameters: HashMap<String, Parameter>,
    parameter_order: Vec<String>,

    bindings: HashMap<String, MeshBinding>,
    binding_order: Vec<String>,

    scene_bindings: HashMap<String, SceneBinding>,
    scene_binding_order: Vec<String>,
    draw_order_groups: Option<Vec<DrawOrderGroup>>,

    blend_key_tables: HashMap<String, BlendShapeKeyTable>,
    blend_key_table_order: Vec<String>,

    blend_constraints: HashMap<String, BlendShapeConstraint>,
    blend_constraint_order: Vec<String>,

    blend_bindings: HashMap<String, BlendShapeBinding>,
    blend_binding_order: Vec<String>,

    glues: HashMap<String, Glue>,
    glue_order: Vec<String>,

    offscreens: HashMap<String, Offscreen>,
    offscreen_order: Vec<String>,

    saved_content: Option<Box<DocumentContent>>,
    lookup: OnceLock<DocumentLookup>,
    prepared: OnceLock<Result<crate::evaluation::PreparedEvaluation, Status>>,
}

#[derive(Debug, Clone, Default)]
pub(super) struct DocumentLookup {
    pub(super) mesh_bindings: HashMap<String, String>,
    pub(super) scene_bindings: HashMap<String, String>,
    pub(super) blend_bindings: HashMap<String, Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq)]
struct DocumentContent {
    id: String,
    canvas: Canvas,
    assets: HashMap<String, ImageAsset>,
    asset_order: Vec<String>,
    parts: HashMap<String, Part>,
    part_order: Vec<String>,
    transforms: HashMap<String, Transform>,
    transform_order: Vec<String>,
    meshes: HashMap<String, Mesh>,
    mesh_order: Vec<String>,
    parameters: HashMap<String, Parameter>,
    parameter_order: Vec<String>,
    bindings: HashMap<String, MeshBinding>,
    binding_order: Vec<String>,
    scene_bindings: HashMap<String, SceneBinding>,
    scene_binding_order: Vec<String>,
    draw_order_groups: Option<Vec<DrawOrderGroup>>,
    blend_key_tables: HashMap<String, BlendShapeKeyTable>,
    blend_key_table_order: Vec<String>,
    blend_constraints: HashMap<String, BlendShapeConstraint>,
    blend_constraint_order: Vec<String>,
    blend_bindings: HashMap<String, BlendShapeBinding>,
    blend_binding_order: Vec<String>,
    glues: HashMap<String, Glue>,
    glue_order: Vec<String>,
    offscreens: HashMap<String, Offscreen>,
    offscreen_order: Vec<String>,
}

#[derive(PartialEq)]
struct ContentRef<'a> {
    id: &'a String,
    canvas: &'a Canvas,
    assets: &'a HashMap<String, ImageAsset>,
    asset_order: &'a Vec<String>,
    parts: &'a HashMap<String, Part>,
    part_order: &'a Vec<String>,
    transforms: &'a HashMap<String, Transform>,
    transform_order: &'a Vec<String>,
    meshes: &'a HashMap<String, Mesh>,
    mesh_order: &'a Vec<String>,
    parameters: &'a HashMap<String, Parameter>,
    parameter_order: &'a Vec<String>,
    bindings: &'a HashMap<String, MeshBinding>,
    binding_order: &'a Vec<String>,
    scene_bindings: &'a HashMap<String, SceneBinding>,
    scene_binding_order: &'a Vec<String>,
    draw_order_groups: &'a Option<Vec<DrawOrderGroup>>,
    blend_key_tables: &'a HashMap<String, BlendShapeKeyTable>,
    blend_key_table_order: &'a Vec<String>,
    blend_constraints: &'a HashMap<String, BlendShapeConstraint>,
    blend_constraint_order: &'a Vec<String>,
    blend_bindings: &'a HashMap<String, BlendShapeBinding>,
    blend_binding_order: &'a Vec<String>,
    glues: &'a HashMap<String, Glue>,
    glue_order: &'a Vec<String>,
    offscreens: &'a HashMap<String, Offscreen>,
    offscreen_order: &'a Vec<String>,
}

impl DocumentContent {
    fn content_ref(&self) -> ContentRef<'_> {
        ContentRef {
            id: &self.id,
            canvas: &self.canvas,
            assets: &self.assets,
            asset_order: &self.asset_order,
            parts: &self.parts,
            part_order: &self.part_order,
            transforms: &self.transforms,
            transform_order: &self.transform_order,
            meshes: &self.meshes,
            mesh_order: &self.mesh_order,
            parameters: &self.parameters,
            parameter_order: &self.parameter_order,
            bindings: &self.bindings,
            binding_order: &self.binding_order,
            scene_bindings: &self.scene_bindings,
            scene_binding_order: &self.scene_binding_order,
            draw_order_groups: &self.draw_order_groups,
            blend_key_tables: &self.blend_key_tables,
            blend_key_table_order: &self.blend_key_table_order,
            blend_constraints: &self.blend_constraints,
            blend_constraint_order: &self.blend_constraint_order,
            blend_bindings: &self.blend_bindings,
            blend_binding_order: &self.blend_binding_order,
            glues: &self.glues,
            glue_order: &self.glue_order,
            offscreens: &self.offscreens,
            offscreen_order: &self.offscreen_order,
        }
    }
}

impl Document {
    pub const SCHEMA_VERSION: u32 = 3;

    pub fn new() -> Self {
        Self::default()
    }

    pub fn initialize(&mut self, id: impl Into<String>, canvas: Canvas) -> Status {
        if self.initialized() {
            return Status::error(
                "ALREADY_INITIALIZED",
                "Create a new Document to open another model.",
            );
        }
        let id_str = id.into();
        if !valid_uuid(&id_str) {
            return Status::error("INVALID_ID", "Use a nonzero canonical lowercase UUID.");
        }
        if !canvas.width.is_finite()
            || !canvas.height.is_finite()
            || canvas.width <= 0.0
            || canvas.height <= 0.0
            || !canvas.origin.x.is_finite()
            || !canvas.origin.y.is_finite()
            || !canvas.pixels_per_unit.is_finite()
            || canvas.pixels_per_unit <= 0.0
        {
            return Status::error(
                "INVALID_CANVAS",
                "Canvas dimensions must be finite and positive.",
            );
        }
        self.id = id_str;
        self.canvas = canvas;
        Status::ok()
    }

    pub fn initialized(&self) -> bool {
        !self.id.is_empty()
    }

    pub fn id(&self) -> &str {
        &self.id
    }

    pub fn canvas(&self) -> Canvas {
        self.canvas
    }

    /// Last document revision that changed visual inputs. Names do not affect frames.
    pub fn evaluation_revision(&self) -> u64 {
        self.evaluation_revision
    }

    pub fn revision(&self) -> u64 {
        self.revision
    }

    fn content_ref(&self) -> ContentRef<'_> {
        ContentRef {
            id: &self.id,
            canvas: &self.canvas,
            assets: &self.assets,
            asset_order: &self.asset_order,
            parts: &self.parts,
            part_order: &self.part_order,
            transforms: &self.transforms,
            transform_order: &self.transform_order,
            meshes: &self.meshes,
            mesh_order: &self.mesh_order,
            parameters: &self.parameters,
            parameter_order: &self.parameter_order,
            bindings: &self.bindings,
            binding_order: &self.binding_order,
            scene_bindings: &self.scene_bindings,
            scene_binding_order: &self.scene_binding_order,
            draw_order_groups: &self.draw_order_groups,
            blend_key_tables: &self.blend_key_tables,
            blend_key_table_order: &self.blend_key_table_order,
            blend_constraints: &self.blend_constraints,
            blend_constraint_order: &self.blend_constraint_order,
            blend_bindings: &self.blend_bindings,
            blend_binding_order: &self.blend_binding_order,
            glues: &self.glues,
            glue_order: &self.glue_order,
            offscreens: &self.offscreens,
            offscreen_order: &self.offscreen_order,
        }
    }

    fn content(&self) -> DocumentContent {
        DocumentContent {
            id: self.id.clone(),
            canvas: self.canvas,
            assets: self.assets.clone(),
            asset_order: self.asset_order.clone(),
            parts: self.parts.clone(),
            part_order: self.part_order.clone(),
            transforms: self.transforms.clone(),
            transform_order: self.transform_order.clone(),
            meshes: self.meshes.clone(),
            mesh_order: self.mesh_order.clone(),
            parameters: self.parameters.clone(),
            parameter_order: self.parameter_order.clone(),
            bindings: self.bindings.clone(),
            binding_order: self.binding_order.clone(),
            scene_bindings: self.scene_bindings.clone(),
            scene_binding_order: self.scene_binding_order.clone(),
            draw_order_groups: self.draw_order_groups.clone(),
            blend_key_tables: self.blend_key_tables.clone(),
            blend_key_table_order: self.blend_key_table_order.clone(),
            blend_constraints: self.blend_constraints.clone(),
            blend_constraint_order: self.blend_constraint_order.clone(),
            blend_bindings: self.blend_bindings.clone(),
            blend_binding_order: self.blend_binding_order.clone(),
            glues: self.glues.clone(),
            glue_order: self.glue_order.clone(),
            offscreens: self.offscreens.clone(),
            offscreen_order: self.offscreen_order.clone(),
        }
    }

    pub fn take_edit_delta(&mut self) -> Option<crate::history::EditDelta> {
        self.receipt.0.take()
    }

    pub(crate) fn remove_unchanged_history_fields(&self, delta: &mut crate::history::EditDelta) {
        use crate::history::{Field, Value};
        delta.values.retain(|key, value| match (key, value) {
            (Field::MeshName(id), Value::Name(name)) => {
                self.get_mesh(id).is_none_or(|m| m.name != *name)
            }
            (Field::Vertex(id, vertex), Value::Position(position)) => {
                self.get_mesh(id).and_then(|m| {
                    self.vertex_slots
                        .get(id)?
                        .get(vertex)
                        .map(|&slot| m.base_positions[slot])
                }) != Some(*position)
            }
            _ => true,
        });
    }

    pub(crate) fn replay_history(&mut self, delta: &mut crate::history::EditDelta) -> EditResult {
        use crate::history::{Field, Value};
        // Validate every target before the first swap; failure cannot partially replay an action.
        for (key, value) in &delta.values {
            let valid = match (key, value) {
                (Field::MeshName(id), Value::Name(_)) => self.meshes.contains_key(id),
                (Field::Vertex(id, vertex), Value::Position(_)) => self
                    .vertex_slots
                    .get(id)
                    .is_some_and(|s| s.contains_key(vertex)),
                _ => false,
            };
            if !valid {
                return self.failed(Status::error(
                    "STALE_HISTORY",
                    "History target is no longer available",
                ));
            }
        }
        self.remove_unchanged_history_fields(delta);
        if delta.values.is_empty() {
            return self.failed(Status::ok());
        }
        let mut meshes = Vec::new();
        let mut positions = false;
        for (key, value) in &mut delta.values {
            match (key, value) {
                (Field::MeshName(id), Value::Name(name)) => {
                    std::mem::swap(&mut self.meshes.get_mut(id).unwrap().name, name);
                    meshes.push(id.clone());
                }
                (Field::Vertex(id, vertex), Value::Position(position)) => {
                    let slot = self.vertex_slots[id][vertex];
                    std::mem::swap(
                        &mut self.meshes.get_mut(id).unwrap().base_positions[slot],
                        position,
                    );
                    positions = true;
                    meshes.push(id.clone());
                }
                _ => unreachable!(),
            }
        }
        meshes.sort();
        meshes.dedup();
        self.changed(
            if positions {
                ChangeKind::Positions
            } else {
                ChangeKind::Metadata
            },
            meshes,
            Vec::new(),
        )
    }

    pub fn same_content(&self, other: &Document) -> bool {
        self.content_ref() == other.content_ref()
    }

    pub fn modified(&self) -> bool {
        match &self.saved_content {
            Some(saved) => self.content_ref() != saved.content_ref(),
            None => self.initialized(),
        }
    }

    pub fn mark_saved(&mut self) {
        self.saved_content = Some(Box::new(self.content()));
    }

    pub fn restore_from(&mut self, source: &Document) {
        let next_rev = self.revision + 1;
        let saved = self.saved_content.clone();
        *self = source.clone();
        self.saved_content = saved;
        self.revision = next_rev;
        self.evaluation_revision = next_rev;
        self.transaction_active = false;
        self.batch_build_active = false;
        self.staged_edits.clear();
    }

    pub fn transaction_active(&self) -> bool {
        self.transaction_active
    }

    /// Begin construction of a new document without publishing a revision and
    /// invalidating derived caches after every inserted object.
    pub fn begin_batch_build(&mut self) -> Status {
        if self.initialized() || self.batch_build_active || self.transaction_active {
            return Status::error(
                "INVALID_BUILD_STATE",
                "Batch construction requires a new document",
            );
        }
        self.batch_build_active = true;
        Status::ok()
    }

    /// Publish a successfully constructed candidate as one document revision.
    pub fn finish_batch_build(&mut self) -> Status {
        if !self.batch_build_active {
            return Status::error("NO_BATCH_BUILD", "Call begin_batch_build first");
        }
        if self.transaction_active || !self.staged_edits.is_empty() {
            return Status::error(
                "INVALID_BUILD_STATE",
                "Commit or cancel the active transaction before finishing the batch",
            );
        }
        if !self.initialized() {
            return Status::error("NOT_INITIALIZED", "Initialize the document first");
        }
        self.batch_build_active = false;
        self.lookup.take();
        self.prepared.take();
        self.receipt.0 = None;
        self.revision = 1;
        self.evaluation_revision = 1;
        Status::ok()
    }

    pub(super) fn mutation_blocked(&self) -> bool {
        self.transaction_active
    }

    pub fn contains_id(&self, id: &str) -> bool {
        self.id == id
            || self.assets.contains_key(id)
            || self.parts.contains_key(id)
            || self.transforms.contains_key(id)
            || self.meshes.contains_key(id)
            || self.parameters.contains_key(id)
            || self.bindings.contains_key(id)
            || self.scene_bindings.contains_key(id)
            || self.blend_key_tables.contains_key(id)
            || self.blend_constraints.contains_key(id)
            || self.blend_bindings.contains_key(id)
            || self.glues.contains_key(id)
            || self.offscreens.contains_key(id)
    }

    pub(super) fn failed(&self, status: Status) -> EditResult {
        EditResult {
            status,
            changes: ChangeSet {
                kind: ChangeKind::None,
                mesh_ids: Vec::new(),
                revision: self.revision,
                object_ids: Vec::new(),
            },
            referrers: Vec::new(),
        }
    }

    pub(crate) fn prepared_evaluation(
        &self,
    ) -> Result<&crate::evaluation::PreparedEvaluation, &Status> {
        self.prepared
            .get_or_init(|| crate::evaluation::PreparedEvaluation::new(self))
            .as_ref()
    }

    pub(super) fn changed(
        &mut self,
        kind: ChangeKind,
        mesh_ids: Vec<String>,
        mut object_ids: Vec<String>,
    ) -> EditResult {
        if self.batch_build_active {
            if object_ids.is_empty() {
                object_ids = mesh_ids.clone();
            }
            if kind == ChangeKind::Structure {
                self.lookup.take();
                self.prepared.take();
            }
            return EditResult {
                status: Status::ok(),
                changes: ChangeSet {
                    kind,
                    mesh_ids,
                    revision: self.revision,
                    object_ids,
                },
                referrers: Vec::new(),
            };
        }
        if object_ids.is_empty() {
            object_ids = mesh_ids.clone();
        }
        // References only change with structure edits; vertex drags retain the index.
        if kind == ChangeKind::Structure {
            self.lookup.take();
            self.prepared.take();
        }
        self.revision += 1;
        if kind != ChangeKind::Metadata {
            self.evaluation_revision = self.revision;
        }
        self.receipt.0 = None;
        EditResult {
            status: Status::ok(),
            changes: ChangeSet {
                kind,
                mesh_ids,
                revision: self.revision,
                object_ids,
            },
            referrers: Vec::new(),
        }
    }

    pub(super) fn lookup(&self) -> &DocumentLookup {
        self.lookup.get_or_init(|| {
            let mut lookup = DocumentLookup::default();
            for id in &self.binding_order {
                lookup
                    .mesh_bindings
                    .insert(self.bindings[id].mesh_id.clone(), id.clone());
            }
            for id in &self.scene_binding_order {
                lookup
                    .scene_bindings
                    .insert(self.scene_bindings[id].target_id().to_owned(), id.clone());
            }
            for id in &self.blend_binding_order {
                if let Some(binding) = self.blend_bindings.get(id) {
                    lookup
                        .blend_bindings
                        .entry(binding.target_id.clone())
                        .or_default()
                        .push(id.clone());
                }
            }
            lookup
        })
    }
}
